### PR TITLE
AVX128: F16C support

### PR DESCRIPTION
**Pull request by Sonicadvance1**
Adds a couple unit tests to ensure MXCSR is adhered to as well

### CODE REVIEW

File: FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp [hunk 1]
```diff
@@ -2607,4 +2610,98 @@ void OpDispatchBuilder::AVX128_VPGATHER(OpcodeArgs) {
...
+    Src = AVX128_LoadSource_WithOpSize(Op, Op->Src[0], Op->Flags, !Is128BitSrc);
+  } else {
+    // In the event that a memory operand is used as the source operand,
+    // the access width will always be half the size of the destination vector width
+    // (i.e. 128-bit vector -> 64-bit mem, 256-bit vector -> 128-bit mem)
+    const auto SrcLoadSize = SrcSize;
+    Src.Low = LoadSource_WithOpSize(FPRClass, Op, Op->Src[0], SrcLoadSize, Op->Flags);
+  }
+
+  auto TransformLow = [this](Ref Src) -> Ref {
```
Comment (alyssarosenzweig) [inline FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp:2630]: These lambdas are frivolous, inline (remove) them.

File: FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp [hunk 2]
```diff
@@ -2607,4 +2610,98 @@ void OpDispatchBuilder::AVX128_VPGATHER(OpcodeArgs) {
...
+      Result.Low = TransformHigh(Result.Low, Src.High);
+    }
+  } else {
+    // No ARM float conversion instructions allow passing in
+    // a rounding mode as an immediate. All of them depend on
+    // the RM field in the FPCR. And so! We have to do some ugly
+    // rounding mode shuffling.
+    const auto NewRMode = Imm8 & 0b11;
+
+    Ref OldRMode = _GetRoundingMode();
```
Comment (alyssarosenzweig) [inline FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp:2683]: Let's land #3766 first, then use those helpers to simplify/optimize this.

File: FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp [hunk 3]
```diff
@@ -2607,4 +2610,98 @@ void OpDispatchBuilder::AVX128_VPGATHER(OpcodeArgs) {
...
+  RefPair Result {};
+  auto TransformLow = [this](Ref Src) -> Ref {
+    return _Vector_FToF(OpSize::i128Bit, OpSize::i16Bit, Src, OpSize::i32Bit);
+  };
+
+  auto TransformHigh = [this](Ref Lower, Ref Src) -> Ref {
+    return _VFCVTN2(OpSize::i128Bit, OpSize::i32Bit, Lower, Src);
+  };
+
+  if (UseMXCSR) {
```
Comment (alyssarosenzweig) [inline FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp:2671]: Might be profitable to split up as

```
Ref OldFPCR;
if (!UseMXCSR) {
   Push()
}

Transform()

if (!Use) {
   Pop()
}
```

Then the transform lambdas are used only once not twice. And then we inline the lambdas and get simpler code overall, I think.

Comment (Sonicadvance1) [reply FEXCore/Source/Interface/Core/OpcodeDispatcher/AVX_128.cpp]: Nice idea

**Comment by Sonicadvance1** [comment]
Passed all f16c unit tests locally since right now the avx128 isn't running unittests in CI.

**Review by lioncash — APPROVED** [review]

**Review by alyssarosenzweig — APPROVED** [review]

**Pull request merged**